File: LeetCode/101-200/114. 二叉树展开为链表.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {void} Do not return anything, modify root in-place instead.
 */
var flatten = function (root) {
  const linkedList = new TreeNode(0)
  let curLinked = linkedList
  const order = (r) => {
    if (r == null) return
    curLinked.right = new TreeNode(r.val)
    curLinked = curLinked.right
    order(r.left)
    order(r.right)
  }
  order(root)
  curLinked = linkedList.right
  let root2 = root
  while (curLinked) {
    root2.val = curLinked.val
    if (curLinked.right) {
      root2.left = null
      root2.right = new TreeNode()
    }
    root2 = root2.right
    curLinked = curLinked.right
  }
}

File: LeetCode/101-200/114. 二叉树展开为链表.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void flatten(TreeNode root) {
        ArrayList<TreeNode> list = new ArrayList<>();
        orderList(root,list);
        int n = list.size();
        for (int i = 1; i < n; i++) {
            TreeNode prev = list.get(i - 1), cur = list.get(i);
            prev.left = null;
            prev.right = cur;
        }
    }

    public void orderList(TreeNode root, ArrayList<TreeNode> list) {
        if (root == null) return ;
        list.add(root);
        orderList(root.left,list);
        orderList(root.right,list);
    }
}